### ISSUE TITLE
Add StateT and trampolining to State documentation

### ISSUE THREAD

**Issue by ceedubs** [non/cats#387]
Documentation for `State` was added in #384. However, it still has a TODO at the end for documenting how `StateT` relates to `State` and how/why trampolining is used for `State`. This TODO should be replaced with documentation.

**Comment by LukaJCB** [comment]
Maybe we should explain the concept of Monad Transformers at some sort of central location? Some of the motivation and explanation already exists in the `OptionT` documentation. We could take some of that and make some sort of Monad Transformer section, or we duplicate some of the documentation, which I'd personally also be okay with. 

Another thought I have, is that cats-mtl is going to be a different package soon AFAIK, so there might also be some duplication there.

**Comment by edmundnoble** [comment]
cats-mtl will provide documentation about monad transformers in general, yes.

**Comment by kailuowang** [comment]
related https://github.com/typelevel/cats/pull/1714
TLDR, we may need a reformulation of `StateT`.
